frontend/src/GameReportForm.tsx

```diff
@@ -65,93 +65,263 @@ const cities: Stronghold[] = [
+    },
+    wasTreebeardMustered: {
+        value: null,
+        error: "",
+        validate: () => null,
+    },
+    usedHandicap: {
+        value: null,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    actionTokens: { value: 0, error: "", validate: () => null },
+    dwarvenRings: { value: 0, error: "", validate: () => null },
+    gameTurns: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    corruption: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    didFellowshipReachMordor: {
+        value: null,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    mordorTrack: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: () => null,
+    },
+    initialEyes: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    wasAragornCrowned: {
+        value: null,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    aragornCrownedTurn: {
+        value: UNINITIALIZED,
+        error: "",
+        validate: () => null,
+    },
+    capturedStrongholds: {
+        value: [],
+        error: "",
+        validate: () => null,
+    },
+    interestRating: {
+        value: -1,
+        error: "",
+        validate: function _() {
+            return detectMissingInput(this.value);
+        },
+    },
+    comment: { value: "", error: "", validate: () => null },
 };
 
 type ValueOf<T> = T[keyof T];
 
 function GameReportForm() {
     const [formData, setFormData] = useState(initialFormData);
+    const [errorOnSubmit, setErrorOnSubmit] = useState<string | null>(null);
 
     const handleInputChange = <K extends keyof FormData>(field: K) => {
-        return (value: FormData[K]) =>
-            setFormData({ ...formData, [field]: value });
+        return (value: FormData[K]["value"]) =>
+            setFormData((prevData) => ({
+                ...prevData,
+                [field]: { ...prevData[field], value },
+            }));
+    };
+
+    const validateField = <K extends keyof FormData>(field: K) => {
+        return () => {
+            setFormData((prevData) => {
+                const error = prevData[field].validate();
+                return error
+                    ? {
+                          ...prevData,
+                          [field]: { ...prevData[field], error },
+                      }
+                    : prevData;
+            });
+        };
     };
 
+    const validateForm = () => {
+        const formError = detectFormErrors(formData);
```

Comment:
I'm not sure what the point is of pre-detecting errors in `detectFormErrors` if we are then going to re-call validate on every field inside the loop? Maybe I'm just missing it.

Reply:
My bad, it's unnecessary, you're not missing anything! It was better in an earlier version but there was a bug and I blamed the wrong thing at first, so I undid it, and forgot I could change it back after I found the real bug - latest version is better.